### DISCUSSION TITLE
List Tables from multiple lines

### DISCUSSION THREAD

**Discussion by gable00** [scandum/tintin#222] [Script help]
Long time lurker first time poster.

I have the following output I would like to capture and out in a list table (small data set example):

**Clan roster for The Emerald Knights - sorted by Player Rank.

No. Name         Current Rank         Lvl Class      Race       Last On
--- ------------ -------------------- --- ---------- ---------- ----------
  1 Aciath       Squire                26 Psionicist Dwarf      01 Feb 2022
 10 Papabear     Squire               108 Warrior    Troll      07 Sep 2007
117 Thindal      Hero                 147 Thief      Human      04 Nov 2021

3 members counted.**

I would like to fire an Action from "Clan roster for The Emerald Knights - sorted by Player Rank." then have each 3 lines with the player information added to a table that I can then manipulate.  Here is what I have worked with so far with no luck. Many thanks!

```
#var data_table {}
#var capturing_data 0

#alias get_roster {
    #send {Roster Emerald 2};
    #var capturing_data 1;
}

#action {^  Clan roster for The Emerald Knights - sorted by Player Rank.$} {
    #if {$capturing_data==1} {
        #nop;
    } else {
        #var capturing_data 1;
    }
} {5.000} {data_capture}

#action {^(\s*\d+)\s+(\S+.*?)\s+(\S+.*?)\s+(\d+)\s+(\S+)\s+(\S+)\s+(\d{1,2}\s+\w{3}\s+\d{4})\s*$} {
    #if {$capturing_data} {
        #var rank $1;
        #var name $2;
        #var title $3;
        #var level $4;
        #var class $5;
        #var race $6;
        #var date $7;
    
        #format row "%d. %s, %s, Level: %d, Class: %s, Race: %s, Date: %s" $rank $name $title $level $class $race $date;
        #list data_table ADD $row;
    }
} {5.000} {data_capture}

#action {^3 members counted.$} {
    #if {$capturing_data} {
        #var capturing_data 0;
    }
} {5.000} {data_capture}

#alias show_table {
    #showme {Full Data Table:};
    #foreach {$data_table} {item} {
        #showme $item;
    };
}
```

**Comment by scandum** [comment]
Hello. Your script had a bunch of errors, but the following should work.

```
#var data_table {}
#var capturing_data 0

#alias get_roster {
    #send {Roster Emerald 2};
    #var capturing_data 1;
}

#action {^  Clan roster for The Emerald Knights - sorted by Player Rank.$} {
    #if {$capturing_data==1} {
        #nop;
    };
    #else {
        #var capturing_data 1;
    }
} {5}

#action {%+1..d %* %* %+1..d %* %* %+1..d %*} {
    #if {$capturing_data} {
        #format row {%+3s. %+12s, %+10s, Level: %+3s, Class: %+12s, Race: %+10s, Date: %s} {%1}{%2}{%3}{%4}{%5}{%6}{%7 %8};
        #list data_table add {$row}
    }
} {5}

#action {^3 members counted.$} {
    #if {$capturing_data} {
        #var capturing_data 0;
    }
} {5}

#alias show_table {
    #showme {Full Data Table:};
    #foreach {$data_table[]} {item} {
        #showme {$item};
    };
}

#alias {test}
{
    #line debug #showme {  Clan roster for The Emerald Knights - sorted by Player Rank.};
    #line debug #showme No. Name Current Rank Lvl Class Race Last On;
    #line debug #showme 1 Aciath Squire 26 Psionicist Dwarf 01 Feb 2022;
    #line debug #showme 10 Papabear Squire 108 Warrior Troll 07 Sep 2007;
    #line debug #showme 117 Thindal Hero 147 Thief Human 04 Nov 2021;
    #line debug #showme 3 members counted.
}
```

**Reply by gable00** [reply]
Very helpful! Thank you.

**Comment by gable00** [comment]
Next question. How would I (or a resource to learn) check the Name from the table we created previously, then match the name on the two actions below and update the table? The "Name" field would be used to match an corresponding updates.

As an example, the first Action below I would want Quests, Total QP and Campains added to the table and in the second Action I would only want the level added to the table.

Aardwolf monthly rankings by Quest Points Earned.

Num    Name           Clan         Level Quests Total Qp Gquests Campaigns
-----  -------------- ----------- ------ ------ -------- ------- ---------
1)     Xenon          Emerald      642    207    22989      14       456
2)     Mithra         Emerald      539    254    18496      13       315
3)     Brontis        Emerald      0      447    16913       8        44
4)     Sylar          Emerald      0      412    16314      13        14
5)     Aleris         Emerald      200    306    14764      11       119
6)     Faelon         Emerald      724    222    14651       3       230
7)     Draven         Emerald      0      503    14284       0         0
8)     Nyla           Emerald      0      439    13469       1         6
9)     Thalia         Emerald      0      439    9906        3        15
10)    Orion          Emerald      0      196    7334        0        17


Aardwolf monthly rankings by Total Levels.

Num    Name         Clan         Level Total Qp Mob Kills Deaths Gquests
-----  ------------ ----------- ------ -------- --------- ------ -------
1)     Azuron       Emerald      724    14651     23352     25       3
2)     Calixte      Emerald      642    22989     11635      1      14
3)     Dervon       Emerald      539    18496     19329     26      13
4)     Elexia       Emerald      230     6479      5219     11       3
5)     Fiora        Emerald      201     4039      6197     43       0
6)     Gryffon      Emerald      200    14764      7072     11      11
7)     Helix        Emerald      173     5103      4092     36      11
8)     Ivarix       Emerald       65     3630      1366      4       0
9)     Jaxon        Emerald       54     3235      2343      0       1
10)    Kyra         Emerald       37     5480     25773      0       0
11)    Lorix        Emerald       33     5744      1712      0       8
12)    Myra         Emerald       15      302       259      2       0
13)    Naxos        Emerald        1      303        27      0       0

Thank yoU!

**Comment by scandum** [comment]
`#help variable` and `#help lists` has some help entries on creating table variables. 

You'd want to do something like this:
```
#action {%+1..d %* %* %+1..d %* %* %+1..d %*} {
    #if {$capturing_data} {
        #format row {%+3s. %+12s, %+10s, Level: %+3s, Class: %+12s, Race: %+10s, Date: %s} {%1}{%2}{%3}{%4}{%5}{%6}{%7 %8};
        #list data_table add {$row};
        #var name_table[%2][level] {%4};
        #var name_table[%2][class] {%5};
        #var name_table[%2][race] {%6}
    }
} {5}
```
Then create some actions to update $name_table for the other commands, and an alias to display $name_table to your liking. For example:
```
#alias {show_table}
{
    #foreach {*name_table[]} {name}
    {
        #echo {Name: %s Level: %s Class: %s} {$name} {$name_table[$name][level]} {$name_table[$name][class]}
    }
}
```

**Comment by gable00** [comment]
Okay, How would I go about updating the data_table to hold the final output? I'm working my way through learning tt++ scripting so apologies if lists are taking me a bit to understand.

We capture the clan roster here:
```
#action {%+1..d %* %* %+1..d %* %* %+1..d %*} {
    #if {$capturing_data} {
        #format row {%+3s. %+12s, %+10s, Level: %+3s, Class: %+12s, Race: %+10s, Date: %s} {%1}{%2}{%3}{%4}{%5}{%6}{%7 %8};
        #list data_table add {$row};
    }
} {5}
```

Then I want to fire another action to get quest point data:
            Aardwolf monthly rankings by Quest Points Earned.

Num    Name         Clan         Level Quests Total Qp Gquests Campaigns
1) Techno       Emerald        659    214    23907      16       473
2) Tipro        Emerald        552    259    18794      13       319
3) Bamboo       Emerald          0    453    17093       8        45

```
#action {^            Aardwolf monthly rankings by Quest Points Earned.$} 
     #if {$capturing_data} {
     capture the data above,
     update the data_table with the quest, total qp, GQ and CP info that matches the Name.
```

How would I append this new data to the data_table and update it with the new information?
    
Thank you!

**Reply by gable00** [reply]
Oh wait. You are saying to get the roster data, then create the name_table and manipulate the name_table with the new action.

Is that right?

For a bit more context I am pulling data from different Actions and using player names as the identifier. This requires multiple Actions to get the data. I'd like to have it all in one table once complete so I can either post it in game or send to a comma delimited file / text file on the server.

**Comment by scandum** [comment]
Yes, you could use:
```
#var name_table[%2][clan] {%3};
```
to add the clan which should be in %3, and the player name should be in %2. Then you need to update the show_table alias to display clans and any other field you want listed.
